src/fde.h

```diff
@@ -52,15 +52,22 @@ class fde
 public:
     fde() {
         *ipaddr = 0;
-        *desc = 0;
         read_handler = nullptr;
         write_handler = nullptr;
         read_method = nullptr;
         write_method = nullptr;
     }
+    // no copying, just move-construct for clear()
+    fde(const fde &) = delete;
+    fde &operator =(const fde &) = delete;
+    fde(fde &&) = default;
```

Comment:
Problem is that statement about being identical is not true. For code such as clear() contains the copy constructor allocates and deallocates a temporary object in addition to the data movement. The trivial move always optimize that case away and just nil-initialize the existing memory area after destructing the data which existed there beforehand. If they *were* identical there would be no need for move operators to have been invented.

Going back to the slow double-allocation to avoid further discussion delays. We can optimize later.

Reply:
> Amos: Problem is that statement about being identical is not true

I am not sute which statement you are referring to. The comment you were responding to does not have the word "identical" AFAICT. I suspect you are talking about this statement:

> Alex: per [reference](https://en.cppreference.com/w/cpp/language/move_constructor): A trivial move constructor is a constructor that performs the same action as the trivial copy constructor, that is, makes a copy of the object representation as if by std::memmove.

That statement is a quote from [reference](https://en.cppreference.com/w/cpp/language/move_constructor). I believe it is true/correct.

> The trivial move always just nil-initialize the existing memory area after destructing the data

The trivial move constructor does none of the above. It does not destruct anything, and it does not nil-initialize any "existing memory area". It only copies old POD bytes into the new POD. It is nothing else but a dumb memmove(). (Note that memmove() does not move anything either.)

> If they were identical there would be no need for move operators to have been invented.

Move operations were invented so that developers can _manually_ optimize interesting/important cases (be writing custom move constructors and assignment operators). They were _not_ invented to _automatically_ optimize moving of PODs (with nil-ifying POD fields and/or destructing something) because either of those actions would break a lot of old code.